### PR TITLE
Sort DFM columns

### PR DESCRIPTION
**Pull request by koheiw**
Sort the columns of DFM to address #2100.

### CODE REVIEW

File: R/dfm.R [hunk 1]
```diff
@@ -226,14 +226,21 @@ dfm.tokens <- function(x,
...
-            meta = attrs[["meta"]]),
-        tolower = FALSE, verbose = verbose
-    )
+
+    temp <- build_dfm(
+        temp, type,
+        docvars = get_docvars(x, user = TRUE, system = TRUE),
+        meta = attrs[["meta"]])
+    
+    if (attrs$meta$object$what != "dictionary") {
```
Comment (koheiw) [inline R/dfm.R:235]: This is to make the order of dictionary keys and column identical. This also skip tokens objects that returned by `tokens_lookup(exclusive = FALSE)`. One solution is not setting what = "dictionary" in non-exclusive mode.

**Comment by koheiw** [comment]
``` r
require(quanteda)
#> Loading required package: quanteda
#> Package version: 3.1.0.9000
#> Unicode version: 13.0
#> ICU version: 66.1
#> Parallel computing: 6 of 6 threads used.
#> See https://quanteda.io for tutorials and examples.
quanteda_options(threads = 8)
#> Warning: Setting threads instead to maximum available 6

toks <- tokens(data_corpus_inaugural)
toks_ng1 <- tokens_ngrams(toks, n = 2)
toks_ng2 <- tokens_ngrams(toks, n = 2)
identical(types(toks_ng1), types(toks_ng2))
#> [1] FALSE
setequal(types(toks_ng1), types(toks_ng2))
#> [1] TRUE

dfmt1 <- dfm(toks_ng1, tolower = FALSE)
dfmt2 <- dfm(toks_ng2, tolower = FALSE)
identical(featnames(dfmt1), featnames(dfmt2))
#> [1] TRUE
```

<sup>Created on 2021-11-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Review by kbenoit — CHANGES_REQUESTED** [review]
OK I'd investigate the answers to these (or implement them) myself but will just comment for now, since I'm still away and not back in London and at my desk properly until tomorrow evening.

I don't understand the full mechanism yet for stabilising the types table and feature orders but it looks like these happen when the "what" attribute is a dictionary. (Clearly I have not gone back through that code - `rebuld_dfm()` and `rebuild_tokens()`, which was not modified in this PR.)

Will this also stabilise the results of `tokens_compound()`? Can you add appropriate tests for that?

Finally for us to resubmit to CRAN pls update `NEWS.md`. (I'll update the version number and the `cran-comments.md`.)

Thanks!

**Comment by koheiw** [comment]
``` r
require(quanteda)
#> Loading required package: quanteda
#> Package version: 3.1.0.9000
#> Unicode version: 13.0
#> ICU version: 66.1
#> Parallel computing: 8 of 10 threads used.
#> See https://quanteda.io for tutorials and examples.
quanteda_options(threads = 8)

toks <- tokens(data_corpus_inaugural)
toks_comp1 <- tokens_compound(toks, phrase("not *"))
toks_comp2 <- tokens_compound(toks, phrase("not *"))
identical(types(toks_comp1), types(toks_comp2))
#> [1] FALSE
setequal(types(toks_comp1), types(toks_comp2))
#> [1] TRUE

dfmt1 <- dfm(toks_comp1, tolower = FALSE)
dfmt2 <- dfm(toks_comp2, tolower = FALSE)
identical(featnames(dfmt1), featnames(dfmt2))
#> [1] TRUE
```

**Comment by koheiw** [comment]
This is just a simple trick. DFM columns are sorted using token types instead of (stochastic) token IDs to make them identical even in multi-threading. If "what = dictionary` in tokens object, columns are in the same order as the dictionary keys.

**Review by kbenoit — APPROVED** [review]

**Pull request merged**